### ISSUE TITLE
Runtime error in Go reading config after updating to 2.4.11

### ISSUE THREAD

**Issue by ptr727** [traefik/traefik#8273]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.containo.us) and didn't find any.

### What did you do?

Watchtower updated my two Traefik instances running on Docker to 2.4.11.

### What did you see instead?

After the update both instances failed with "Error in Go routine: runtime error: invalid memory address or nil pointer dereference".

### What version of Traefik are you using?

2.4.9 works, 2.4.11 fails.

### What is your environment & configuration?

traefik.yml:
```yaml
global:
  sendAnonymousUsage: true

log:
  level: "WARN" # DEBUG, PANIC, FATAL, ERROR, WARN, and INFO

api:
  dashboard: true
  insecure: true # Enable dashboard on http 8080 for direct access

serversTransport:
  insecureSkipVerify: true # Ignore downstream SSL cert errors
  
providers:
  docker:
    exposedByDefault: false
    network: "localnet" # {{ docker_local_network }}
  file:
    filename: "/config/dynamic.yml"

# https://docs.traefik.io/routing/entrypoints/
entryPoints:
  web:
    address: ":80"
    http:
      redirections:
        entryPoint:
          to: websecure
          scheme: https
  websecure:
    address: ":443"
    http:
      # --entrypoints.websecure.http.tls=true
      # --entrypoints.websecure.http.tls.certresolver=dns-cloudflare
      # --certificatesresolvers.dns-cloudflare.acme.email={{ cloudflare_email }}
      # --certificatesresolvers.dns-cloudflare.acme.storage=/config/acme.json
      # --certificatesresolvers.dns-cloudflare.acme.dnschallenge.provider=cloudflare
      # CF_DNS_API_TOKEN: "{{ cloudflare_dns_api_token }}"
      tls:
        domains:
          - main: "home.insanegenius.net" # {{ ansible_domain }}
            sans:
              - "*.home.insanegenius.net" #*.{{ ansible_domain }}
          - main: "insanegenius.net" # {{ external_domain }}
            sans:
              - "*.insanegenius.net" # *.{{ external_domain }}
```

dynamic.yml:
```yaml
tls:
  stores:
    default: {}
  certificates:
    - certFile: "/config/star_insanegenius_net.crt"
      keyFile: "/config/star_insanegenius_net.key"
    - certFile: "/config/star_home_insanegenius_net.crt"
      keyFile: "/config/star_home_insanegenius_net.key"
```

### If applicable, please paste the log output in DEBUG level

```
time="2021-07-16T02:00:37-07:00" level=error msg="Error in Go routine: runtime error: invalid memory address or nil pointer dereference",time="2021-07-16T02:00:37-07:00" level=info msg="Configuration loaded from file: /config/traefik.yml",
time="2021-07-16T02:00:37-07:00" level=error msg="Stack: goroutine 85 [running]:\nruntime/debug.Stack(0xc000194000, 0x343c2ba, 0x17)\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x9f\ngithub.com/traefik/traefik/v2/pkg/safe.defaultRecoverGoroutine(0x2e89ec0, 0x546faf0)\n\t/go/src/github.com/traefik/traefik/pkg/safe/routine.go:66 +0xb4\ngithub.com/traefik/traefik/v2/pkg/safe.GoWithRecover.func1.1(0x3557b00)\n\t/go/src/github.com/traefik/traefik/pkg/safe/routine.go:56 +0x5a\npanic(0x2e89ec0, 0x546faf0)\n\t/usr/local/go/src/runtime/panic.go:965 +0x1b9\ngithub.com/traefik/traefik/v2/pkg/provider/file.(*Provider).loadFileConfig(0xc0005a94d0, 0x3999d80, 0xc0008a5b00, 0xc000057068, 0x13, 0x3999d01, 0xc0008a5b00, 0x18, 0x18)\n\t/go/src/github.com/traefik/traefik/pkg/provider/file/file.go:194 +0xb20\ngithub.com/traefik/traefik/v2/pkg/provider/file.(*Provider).BuildConfiguration(0xc0005a94d0, 0x18, 0x7faf07565f18, 0x18)\n\t/go/src/github.com/traefik/traefik/pkg/provider/file/file.go:85 +0xf8\ngithub.com/traefik/traefik/v2/pkg/provider/file.(*Provider).Provide(0xc0005a94d0, 0xc000590780, 0xc0008a4c90, 0xc000594f20, 0x2)\n\t/go/src/github.com/traefik/traefik/pkg/provider/file/file.go:49 +0x32\ngithub.com/traefik/traefik/v2/pkg/provider/aggregator.launchProvider(0xc000590780, 0xc0008a4c90, 0x3959690, 0xc0005a94d0)\n\t/go/src/github.com/traefik/traefik/pkg/provider/aggregator/aggregator.go:151 +0x1c2\ngithub.com/traefik/traefik/v2/pkg/provider/aggregator.ProviderAggregator.Provide(0x3959de8, 0xc0003d7f80, 0x3959690, 0xc0005a94d0, 0xc000594a80, 0x2, 0x2, 0xc000590780, 0xc0008a4c90, 0x0, ...)\n\t/go/src/github.com/traefik/traefik/pkg/provider/aggregator/aggregator.go:123 +0x1bb\ngithub.com/traefik/traefik/v2/pkg/server.(*ConfigurationWatcher).startProvider.func1()\n\t/go/src/github.com/traefik/traefik/pkg/server/configurationwatcher.go:96 +0x68\ngithub.com/traefik/traefik/v2/pkg/safe.GoWithRecover.func1(0x3557b00, 0xc0008a5a10)\n\t/go/src/github.com/traefik/traefik/pkg/safe/routine.go:59 +0x4f\ncreated by github.com/traefik/traefik/v2/pkg/safe.GoWithRecover\n\t/go/src/github.com/traefik/traefik/pkg/safe/routine.go:53 +0x49\n"
```
```
Stack: goroutine 85 [running]:
runtime/debug.Stack(0xc000194000, 0x343c2ba, 0x17)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
github.com/traefik/traefik/v2/pkg/safe.defaultRecoverGoroutine(0x2e89ec0, 0x546faf0)
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:66 +0xb4
github.com/traefik/traefik/v2/pkg/safe.GoWithRecover.func1.1(0x3557b00)
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:56 +0x5a
panic(0x2e89ec0, 0x546faf0)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/traefik/traefik/v2/pkg/provider/file.(*Provider).loadFileConfig(0xc0005a94d0, 0x3999d80, 0xc0008a5b00, 0xc000057068, 0x13, 0x3999d01, 0xc0008a5b00, 0x18, 0x18)
	/go/src/github.com/traefik/traefik/pkg/provider/file/file.go:194 +0xb20
github.com/traefik/traefik/v2/pkg/provider/file.(*Provider).BuildConfiguration(0xc0005a94d0, 0x18, 0x7faf07565f18, 0x18)
	/go/src/github.com/traefik/traefik/pkg/provider/file/file.go:85 +0xf8
github.com/traefik/traefik/v2/pkg/provider/file.(*Provider).Provide(0xc0005a94d0, 0xc000590780, 0xc0008a4c90, 0xc000594f20, 0x2)
	/go/src/github.com/traefik/traefik/pkg/provider/file/file.go:49 +0x32
github.com/traefik/traefik/v2/pkg/provider/aggregator.launchProvider(0xc000590780, 0xc0008a4c90, 0x3959690, 0xc0005a94d0)
	/go/src/github.com/traefik/traefik/pkg/provider/aggregator/aggregator.go:151 +0x1c2
github.com/traefik/traefik/v2/pkg/provider/aggregator.ProviderAggregator.Provide(0x3959de8, 0xc0003d7f80, 0x3959690, 0xc0005a94d0, 0xc000594a80, 0x2, 0x2, 0xc000590780, 0xc0008a4c90, 0x0, ...)
	/go/src/github.com/traefik/traefik/pkg/provider/aggregator/aggregator.go:123 +0x1bb
github.com/traefik/traefik/v2/pkg/server.(*ConfigurationWatcher).startProvider.func1()
	/go/src/github.com/traefik/traefik/pkg/server/configurationwatcher.go:96 +0x68
github.com/traefik/traefik/v2/pkg/safe.GoWithRecover.func1(0x3557b00, 0xc0008a5a10)
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:59 +0x4f
created by github.com/traefik/traefik/v2/pkg/safe.GoWithRecover
	/go/src/github.com/traefik/traefik/pkg/safe/routine.go:53 +0x49
```

**Comment by dtomcej** [comment]
Hello @ptr727,

if you remove the empty store:
```
tls:
  certificates:
    - certFile: "/config/star_insanegenius_net.crt"
      keyFile: "/config/star_insanegenius_net.key"
    - certFile: "/config/star_home_insanegenius_net.crt"
      keyFile: "/config/star_home_insanegenius_net.key"
```

Does it work as expected?

**Comment by ptr727** [comment]
I removed the default empty store, no more error :)

I added the empty config based on docs when I created the config.
But looking at the docs now the format did seem to change: https://doc.traefik.io/traefik/https/tls/

I do now get these TLS warnings for all my services:
```
time="2021-07-16T10:57:52-07:00" level=warning msg="No domain found in rule PathPrefix(`/whoami`) || HostHeader(`whoami.home.insanegenius.net`) || HostHeader(`whoami.insanegenius.net`), the TLS options applied for this router will depend on the hostSNI of each request" entryPointName=websecure routerName=whoami@docker
```

Ansible docker config for whoami:
```
    labels:
      traefik.enable: "true"
      traefik.http.routers.whoami.rule: "PathPrefix(`/whoami`) || HostHeader(`whoami.{{ ansible_domain }}`) || HostHeader(`whoami.{{ external_domain }}`)"
      traefik.http.routers.whoami.entrypoints: "websecure"
```

Not sure when these warning started, they were not there when I created the config, maybe they came later, or maybe they are related to this TLS store change?

**Comment by dtomcej** [comment]
@ptr727

The `HostHeader` rule was implemented to help deal with domain fronting, which may not be what you intend. If you are not domain fronting, then you should be able to replace those rules with `Host` rules, which will remove that warning.

**Comment by ptr727** [comment]
Hmm, with v2.2 Host() stopped working, I remember because after the update to overnight update to 2.2 nothing routed the next morning, and there were lots of reports of the break on forums and reddit. 

If I recall correctly this was a after a "fix" to prevent domain fronting:
https://github.com/traefik/traefik/pull/7008
https://github.com/traefik/traefik/issues/7020
The "fix" was to switch from using Host() to use HostHeader().
I can see from my config history that I made this change 9 July 2020, comment "fix for Traefik v2.2".

As of now the documentation for Host() and HostHeader() are exactly the same?
https://doc.traefik.io/traefik/routing/routers/#rule
```
Host(`example.com`, ...) | Check if the request domain (host header value) targets one of the given domains.
HostHeader(`example.com`, ...) | Check if the request domain (host header value) targets one of the given domains.
```

So, I know I'm going on a tangent, but I still wonder, do let me know if I should move my question to the forum?

**Comment by dtomcej** [comment]
@ptr727,

You are correct, it was implemented in 2.2.2, and then effectively removed in 2.2.5.

The code for the Host and HostHeader rules call the exact same method:

https://github.com/traefik/traefik/blob/v2.4/pkg/rules/rules.go#L16-17

However some code such as the ACME client, and TLS code that only check `Host` rules, since they were written pre-2.2.2 and post-2.2.5.

**Comment by ptr727** [comment]
Ok, reverted all HostHeader() back to Host(), and no more warnings, thank you.

**Comment by traefiker** [comment]
Closed by #8274.
